src/transformers/models/univnet/feature_extraction_univnet.py

```diff
@@ -0,0 +1,466 @@
+        self.normalize_min = normalize_min
+        self.normalize_max = normalize_max
+        self.model_in_channels = model_in_channels
+        self.pad_end_length = pad_end_length
+        # TODO: Calculate spectrogram_zero? Should be a function of self.mel_floor I think
+        self.spectrogram_zero = spectrogram_zero
+
+    # Based on tacotron2.audio_processing.dynamic_range_compression
+    # https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L78
+    def dynamic_range_compression(self, waveform: np.ndarray):
+        return np.log(np.clip(waveform, a_min=self.compression_clip_val, a_max=None) * self.compression_factor)
+
+    # Based on tacotron2.audio_processing.dynamic_range_compression
+    # https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L87
+    def dynamic_range_decompression(self, waveform: np.ndarray):
+        return np.exp(waveform) / self.compression_factor
+
+    def normalize(self, spectrogram):
+        return 2 * ((spectrogram - self.normalize_min) / (self.normalize_max - self.normalize_min)) - 1
+
+    def denormalize(self, spectrogram):
+        return self.normalize_min + (self.normalize_max - self.normalize_min) * ((spectrogram + 1) / 2)
+
+    def mel_spectrogram(self, waveform: np.ndarray) -> np.ndarray:
+        """
+        Calculates log MEL spectrograms from a batch of waveforms.
+
+        Args:
+            waveform (`np.ndarray` of shape `(length,)`):
+                The input waveform. This must be a single real-valued, mono waveform.
+
+        Returns:
+            `numpy.ndarray` containing a spectrogram of shape `(num_frames, num_mel_bins)`.
+        """
+        # Do custom padding
+        # See https://github.com/maum-ai/univnet/blob/master/utils/stft.py#L84
+        waveform = np.pad(
+            waveform,
+            (int((self.n_fft - self.hop_length) / 2), int((self.n_fft - self.hop_length) / 2)),
+            mode="reflect",
+        )
+
+        # Get the complex spectrogram.
+        # Note: waveform must be unbatched currently due to the implementation of spectrogram(...).
+        complex_spectrogram = spectrogram(
+            waveform,
+            window=self.window,
+            frame_length=self.n_fft,
+            hop_length=self.hop_length,
+            fft_length=self.n_fft,
+            power=None,
+            center=self.center,
+            mel_filters=None,
+            mel_floor=None,
+        )
+
+        # Apply the MEL filter bank and MEL floor manually since UnivNet uses a slightly different implementation
+        amplitude_spectrogram = np.sqrt(
+            np.real(complex_spectrogram) ** 2 + np.imag(complex_spectrogram) ** 2 + self.mel_floor
+        )
+        mel_spectrogram = np.matmul(self.mel_filters.T, amplitude_spectrogram)
+
+        # Perform spectral normalization to get the log mel spectrogram.
+        log_mel_spectrogram = self.dynamic_range_compression(mel_spectrogram)
```

Comment:
My understanding is that this step is necessary - because the [UnivNet spectrogram implementation](https://github.com/maum-ai/univnet/blob/master/utils/stft.py#L71) is different from the [`transformers.audio_utils.spectrogram`](https://github.com/huggingface/transformers/blob/main/src/transformers/audio_utils.py#L267) implementation (in particular, in the way `mel_floor` is used when converting a complex spectrogram to an amplitude spectrogram), I apply the mel filter bank and convert to a log mel spectrogram using `self.dynamic_range_compression` manually in `UnivNetFeatureExtractor.mel_spectrogram`.

`self.dynamic_range_compression` is similar to [`transformers.audio_utils.power_to_db`](https://github.com/huggingface/transformers/blob/main/src/transformers/audio_utils.py#L455) and [`transformers.audio_utils.amplitude_to_db`](https://github.com/huggingface/transformers/blob/main/src/transformers/audio_utils.py#L506), with the differences being that `self.dynamic_range_compression` uses the natural log instead of log base 10 and it doesn't have a factor (e.g. `10.0`/`20.0`) in front of the log. Unless I'm missing something, that means it's not easy to reimplement `self.dynamic_range_compression` using `transformers.audio_utils.power_to_db` or `transformers.audio_utils.amplitude_to_db`.

Reply:
I think what @ArthurZucker is suggesting is that there's no need to define `dynamic_range_compression` as a one-line function. Instead, it would be cleaner to do all the operations in the main `mel_spectrogram` function:
```suggestion
log_mel_spectrogram =  np.log(np.clip(mel_spectrogram, a_min=self.compression_clip_val, a_max=None) * self.compression_factor)
```

Reply:
I have moved the `dynamic_range_compression` calculation into `mel_spectrogram`.